src/capwap-dp.c

```diff
@@ -890,6 +914,165 @@ static void erl_get_stats(int arity, ei_x_buff *x_in, ei_x_buff *x_out)
+{
+    unsigned long sum = 0;
+    const uint16_t *ip1;
+
+    ip1 = ptr;
+    while (count > 1)
+    {
+        sum += *ip1++;
+        if (sum & 0x80000000)
+            sum = (sum & 0xFFFF) + (sum >> 16);
+        count -= 2;
+    }
+
+    while (sum >> 16)
+        sum = (sum & 0xFFFF) + (sum >> 16);
+
+    return (~sum);
+}
+
+static struct ether_header* fill_raw_udp_packet(void *send_data,
+        uint16_t data_len, struct in_addr daddr, uint8_t *mac_dhost, uint16_t *send_len)
+{
+    char *datagram, *data, *pseudogram;
+    uint16_t ip_tot_len, tot_len;
+    struct ifreq if_mac, if_ip;
+
+    ip_tot_len = data_len + sizeof(struct iphdr) + sizeof(struct udphdr);
+    tot_len = ip_tot_len + sizeof(struct ether_header);
+    *send_len = tot_len;
+
+    datagram = (char *) calloc(1, tot_len);
+
+    struct udp_pheader psh;
+
+    struct ether_header *eh = (struct ether_header *) datagram;
+
+    struct iphdr *iph = (struct iphdr *) (datagram +
+            sizeof(struct ether_header));
+    struct udphdr *udph = (struct udphdr *) (datagram +
+            sizeof(struct ether_header) + sizeof(struct iphdr));
+
+    data = datagram + sizeof(struct ether_header) + sizeof(struct iphdr) +
+        sizeof(struct udphdr);
+    memcpy(data, send_data, data_len);
+
+    // Ethernet frame
+    // Get the MAC address of the TAP interface
+    memset(&if_mac, 0, sizeof(struct ifreq));
+    strncpy(if_mac.ifr_name, tap_dev, IFNAMSIZ);
+    if (ioctl(workers[send_worker].tap_fd, SIOCGIFHWADDR, &if_mac) < 0) {
+        perror("SIOCGIFHWADDR");
+    }
+    eh->ether_type = htons(ETH_P_IP);
+    memcpy(eh->ether_shost, if_mac.ifr_hwaddr.sa_data, 6);
+    memcpy(eh->ether_dhost, mac_dhost, 6);
+
+
+    // IP Header
+    // Get the IP address of the TAP interface
+    memset(&if_ip, 0, sizeof(struct ifreq));
+    strncpy(if_ip.ifr_name, tap_dev, IFNAMSIZ);
+    if (ioctl(workers[send_worker].tap_fd, SIOCGIFADDR, &if_ip) < 0) {
+        perror("SIOCGIFADDR");
+    }
+    char *tap_ip = inet_ntoa(( (struct sockaddr_in *)&if_ip.ifr_addr )->sin_addr);
+
+    iph->ihl = 5;
+    iph->version = 4;
+    iph->tot_len = htons(ip_tot_len);
+    iph->id = htons(0);
+    iph->frag_off = 0;
+    iph->ttl = 255;
+    iph->protocol = IPPROTO_UDP;
+    // Source ip
+    iph->saddr = inet_addr(tap_ip);
+    // Dest ip
+    iph->daddr = daddr.s_addr;
+
+    // Ip checksum
+    iph->check = csum((unsigned short *)iph, sizeof(struct iphdr));
+
+    //UDP header
+    udph->source = htons(67);
+    udph->dest = htons(68);
+    udph->len = htons(sizeof(struct udphdr) + data_len);
+
+    //Now the UDP checksum using the pseudo header
+    psh.source_address = iph->saddr;
+    psh.dest_address = daddr.s_addr;
+    psh.placeholder = 0;
+    psh.protocol = IPPROTO_UDP;
+    psh.udp_length = htons(sizeof(struct udphdr) + data_len);
+
+    int psize = sizeof(struct udp_pheader) + sizeof(struct udphdr) + data_len;
+    pseudogram = malloc(psize);
```

Comment:
no way, the checksum can be calculated incrementally without building a linear copy of the packet.